silx/opencl/sparse.py

```diff
@@ -0,0 +1,319 @@
+            setattr(self, arr_name, self.cl_mem[arr_name])
+            self.cl_mem[arr_name].fill(0) # allocate_buffers() uses empty()
+        self._old_array = self.array
+        self._old_data = self.data
+        self._old_indices = self.indices
+        self._old_indptr = self.indptr
+
+
+    def _setup_kernels(self):
+        self._setup_compaction_kernel()
+        self._setup_decompaction_kernel()
+
+
+    def _setup_compaction_kernel(self):
+        self.scan_kernel = GenericScanKernel(
+            self.ctx, self.indice_dtype,
+            arguments="__global float* data, __global float *data_compacted, __global int *indices, __global int* indptr",
+            input_expr="(fabs(data[i]) > 0.0f) ? 1 : 0",
+            scan_expr="a+b", neutral="0",
+            output_statement="""
+                // item is the running sum of input_expr(i), i.e the cumsum of "nonzero"
+                if (prev_item != item) {
+                    data_compacted[item-1] = data[i];
+                    indices[item-1] = GET_INDEX(i);
+                }
+                // The last cumsum element of each line of "nonzero" goes to inptr[i]
+                if ((i+1) % IMAGE_WIDTH == 0) {
+                    indptr[(i/IMAGE_WIDTH)+1] = item;
+                }
+                """,
+            options="-DIMAGE_WIDTH=%d" % self.shape[1],
+            preamble="#define GET_INDEX(i) (i % IMAGE_WIDTH)",
+        )
+
+
+    def _setup_decompaction_kernel(self):
+        OpenclProcessing.compile_kernels(
+            self,
+            self.kernel_files,
+            compile_options=["-DIMAGE_WIDTH=%d" % self.shape[1]]
+        )
+        device = self.ctx.devices[0]
+        wg_x = min(
+            device.max_work_group_size,
+            32,
+            self.kernels.max_workgroup_size("densify_csr")
+        )
+        self._decomp_wg = (wg_x, 1)
+        self._decomp_grid = (self._decomp_wg[0], self.shape[0])
+
+
+    # --------------------------------------------------------------------------
+    # -------------------------- Array utils -----------------------------------
+    # --------------------------------------------------------------------------
+
+    # TODO handle pyopencl Buffer
+    def check_array(self, arr):
+        """
+        Check that provided array is compatible with current context.
+
+        :param arr: numpy.ndarray or pyopencl.array.Array
+            2D array in dense format.
+        """
+        assert arr.size == self.size
+        assert arr.dtype == np.float32
+
+
+    # TODO handle pyopencl Buffer
+    def check_sparse_arrays(self, arrays):
+        """
+        Check that the provided sparse arrays are compatible with the current
+        context.
+
+        :param arrays: tuple of numpy.ndarray or pyopencl.array.Array
+            Tuple of length 3. It contains the arrays "data", "indices", "indptr"
+        """
+        assert len(arrays) == 3
+        data, ind, indptr = arrays
```

Comment:
What about using a namedtuple for this ?

Reply:
In the API, I would prefer using simple data structures for user convenience. But we might use named tuple in the internal mechanisms - is this what you suggest ?